File: mobile/ios/Virtu/ShellBridge.swift
import UIKit
import WebKit

/// The native half of bridge protocol v1 (client/src/shell.md): a
/// `WKScriptMessageHandlerWithReply` in the page content world, plus the
/// document-start shim that presents the uniform `window.virtuShell`.
/// Message parsing/serialization lives in `ShellProtocol`; this class is
/// only the WebKit/UIKit glue.
final class ShellBridge: NSObject, WKScriptMessageHandlerWithReply {
  static let handlerName = "virtuShell"

  private let keychain: KeychainStore
  private weak var presenter: UIViewController?
  private weak var userContentController: WKUserContentController?
  private let shimSource: String?

  init(keychain: KeychainStore, presenter: UIViewController) {
    self.keychain = keychain
    self.presenter = presenter
    let version =
      (Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String) ?? "0"
    if let url = Bundle.main.url(forResource: "shell-bridge", withExtension: "js"),
       let source = try? String(contentsOf: url, encoding: .utf8) {
      shimSource = source.replacingOccurrences(of: "__SHELL_VERSION__", with: version)
    } else {
      // Missing resource: inject nothing — the SPA runs as plain web, the
      // degradation shell.md promises.
      shimSource = nil
    }
    super.init()
  }

  func attach(to controller: WKUserContentController) {
    userContentController = controller
    controller.addScriptMessageHandler(self, contentWorld: .page, name: Self.handlerName)
    refreshUserScripts()
  }

  func userContentController(
    _ userContentController: WKUserContentController,
    didReceive message: WKScriptMessage,
    replyHandler: @escaping (Any?, String?) -> Void
  ) {
    // Never reject the JS promise (the second closure argument): shell.md's
    // contract is that errors come back as error replies.
    guard let body = message.body as? String else {
      replyHandler(ShellProtocol.errorReply("bad-payload"), nil)
      return
    }
    switch ShellProtocol.parse(body) {
    case .malformed(let error):
      replyHandler(ShellProtocol.errorReply(error), nil)
    case .request(let command):
      execute(command) { ok in
        replyHandler(ok ? ShellProtocol.okReply() : ShellProtocol.errorReply("failed"), nil)
      }
    }
  }

  private func execute(_ command: ShellProtocol.Command, completion: @escaping (Bool) -> Void) {
    switch command {
    case .storeApiKey(let key):
      let ok = keychain.store(apiKey: key)
      refreshUserScripts()
      completion(ok)

    case .clearApiKey:
      let ok = keychain.clear()
      refreshUserScripts()
      completion(ok)

    case .share(let title, let text, let url):
      guard let presenter else { return completion(false) }
      var items: [Any] = []
      if let title { items.append(title) }
      if let text { items.append(text) }
      if let url, let shareURL = URL(string: url) { items.append(shareURL) }
      let sheet = UIActivityViewController(activityItems: items, applicationActivities: nil)
      // iPad requires a popover anchor or UIKit crashes on present.
      sheet.popoverPresentationController?.sourceView = presenter.view
      sheet.popoverPresentationController?.sourceRect = CGRect(
        x: presenter.view.bounds.midX, y: presenter.view.bounds.midY, width: 0, height: 0
      )
      presenter.present(sheet, animated: true)
      // ok = "sheet presented"; completion/cancel deliberately unreported —
      // the protocol keeps to Android's common denominator (shell.md).
      completion(true)

    case .openExternal(let url):
      UIApplication.shared.open(url, options: [:]) { success in completion(success) }
    }
  }

  /// The localStorage safety net (plans/mobile.md, Auth): WKWebView storage
  /// has documented flakiness across suspensions, so if it was evicted while
  /// the Keychain still holds the key, re-seed at document start instead of
  /// forcing a re-login. Re-registered on every store/clear; takes effect on
  /// the next (re)load, which is exactly when eviction would bite.
  private func refreshUserScripts() {
    guard let controller = userContentController else { return }
    controller.removeAllUserScripts()
    if let shimSource {
      controller.addUserScript(
        WKUserScript(source: shimSource, injectionTime: .atDocumentStart, forMainFrameOnly: true, in: .page)
      )
    }
    if let key = keychain.load(), let literal = Self.jsStringLiteral(key) {
      let healing =
        #"if (!localStorage.getItem("virtu.apiKey")) localStorage.setItem("virtu.apiKey", \#(literal));"#
      controller.addUserScript(
        WKUserScript(source: healing, injectionTime: .atDocumentStart, forMainFrameOnly: true, in: .page)
      )
    }
  }

  /// A JSON string literal is also a valid JS string literal — safe
  /// interpolation into the healing script.
  private static func jsStringLiteral(_ value: String) -> String? {
    guard let data = try? JSONSerialization.data(withJSONObject: value, options: .fragmentsAllowed)
    else { return nil }
    return String(data: data, encoding: .utf8)
  }
}

File: mobile/android/app/src/main/java/email/zinc/virtu/ShellBridge.kt
package email.zinc.virtu

import android.app.Activity
import android.content.ActivityNotFoundException
import android.content.Intent
import android.net.Uri
import android.webkit.WebView
import androidx.webkit.JavaScriptReplyProxy
import androidx.webkit.ScriptHandler
import androidx.webkit.WebViewCompat
import androidx.webkit.WebViewFeature
import org.json.JSONObject

/**
 * The native half of bridge protocol v1 (client/src/shell.md): receives the
 * web app's messages over an origin-allowlisted WebMessageListener (never
 * `addJavascriptInterface` — plans/mobile.md, platform gotchas) and injects
 * the document-start shim that presents the uniform `window.virtuShell`.
 * Message parsing/serialization lives in [ShellProtocol]; this class is only
 * the Android glue.
 */
class ShellBridge(
  private val activity: Activity,
  private val apiKeyStore: ApiKeyStore,
  private val webOrigin: String,
) {
  private var webView: WebView? = null
  private var healingScript: ScriptHandler? = null

  /**
   * Returns false when the device's WebView predates the required features
   * (rare on minSdk 26+, but possible with a never-updated system WebView).
   * Nothing is injected then and the SPA behaves as a plain browser tab —
   * exactly the degradation shell.md promises.
   */
  fun attach(webView: WebView): Boolean {
    if (
      !WebViewFeature.isFeatureSupported(WebViewFeature.WEB_MESSAGE_LISTENER) ||
      !WebViewFeature.isFeatureSupported(WebViewFeature.DOCUMENT_START_SCRIPT)
    ) {
      return false
    }
    this.webView = webView

    WebViewCompat.addWebMessageListener(webView, PORT_NAME, setOf(webOrigin)) {
        _, message, _, isMainFrame, replyProxy ->
      val data = message.data
      if (data != null) handle(data, replyProxy, isMainFrame)
    }

    val shim = activity.assets.open("shell-bridge.js").bufferedReader().use { it.readText() }
      .replace("__SHELL_VERSION__", BuildConfig.VERSION_NAME)
    WebViewCompat.addDocumentStartJavaScript(webView, shim, setOf(webOrigin))

    refreshHealingScript()
    return true
  }

  private fun handle(envelopeJson: String, replyProxy: JavaScriptReplyProxy, isMainFrame: Boolean) {
    when (val inbound = ShellProtocol.parse(envelopeJson)) {
      is ShellProtocol.Inbound.Malformed -> {
        // A null id means the envelope itself was unreadable — no reply to
        // correlate, so drop it (only our own shim produces envelopes; the
        // seam's reply timeout covers the impossible case).
        val id = inbound.id ?: return
        replyProxy.postMessage(ShellProtocol.errorReply(id, inbound.error))
      }
      is ShellProtocol.Inbound.Request -> {
        // Only the main frame drives native behavior; a same-origin subframe
        // gets an error reply, never a silent drop — every dropped reply is a
        // web-side promise that hangs until the seam's timeout.
        val ok = isMainFrame && runCatching { execute(inbound.command) }.getOrDefault(false)
        replyProxy.postMessage(
          if (ok) ShellProtocol.okReply(inbound.id)
          else ShellProtocol.errorReply(inbound.id, "failed"),
        )
      }
    }
  }

  private fun execute(command: ShellProtocol.Command): Boolean = when (command) {
    is ShellProtocol.Command.StoreApiKey -> {
      apiKeyStore.store(command.key)
      refreshHealingScript()
      true
    }
    is ShellProtocol.Command.ClearApiKey -> {
      apiKeyStore.clear()
      refreshHealingScript()
      true
    }
    is ShellProtocol.Command.Share -> {
      val send = Intent(Intent.ACTION_SEND).setType("text/plain")
      command.title?.let { send.putExtra(Intent.EXTRA_SUBJECT, it) }
      send.putExtra(Intent.EXTRA_TEXT, listOfNotNull(command.text, command.url).joinToString("\n"))
      activity.startActivity(Intent.createChooser(send, null))
      // ok = "sheet presented"; the protocol deliberately does not report
      // whether the user completed the share (shell.md).
      true
    }
    is ShellProtocol.Command.OpenExternal -> try {
      activity.startActivity(
        Intent(Intent.ACTION_VIEW, Uri.parse(command.url)).addCategory(Intent.CATEGORY_BROWSABLE),
      )
      true
    } catch (e: ActivityNotFoundException) {
      false
    }
  }

  /**
   * The localStorage safety net (plans/mobile.md, Auth): if the WebView's
   * storage is ever evicted while Keystore still holds the key, re-seed it at
   * document start instead of forcing a re-login. Re-registered on every
   * store/clear so it always reflects the current key; takes effect on the
   * next (re)load, which is exactly when eviction would bite.
   */
  private fun refreshHealingScript() {
    val target = webView ?: return
    healingScript?.remove()
    healingScript = null
    val key = apiKeyStore.load() ?: return
    // JSONObject.quote emits a JSON string literal, which is also a valid JS
    // string literal — safe interpolation into the script.
    val script =
      """if (!localStorage.getItem("virtu.apiKey")) localStorage.setItem("virtu.apiKey", ${JSONObject.quote(key)});"""
    healingScript = WebViewCompat.addDocumentStartJavaScript(target, script, setOf(webOrigin))
  }

  private companion object {
    /** The raw port object the shim wraps; the web app never touches it. */
    const val PORT_NAME = "virtuShellPort"
  }
}
